Fix randomint(), randomint_seeded(), randombool(), and randombool_seeded()

MasonM opened the pull request:
The [rand.Int63n(n)](https://pkg.go.dev/math/rand#Rand.Int63n) function returns an integer in the range `[0, n)`, but [randomint(from, to)](https://developer.fastly.com/reference/vcl/functions/randomness/randomint/) returns an integer in the range `[from, to]`. Therefore, we need to add one to `n` to get the full range.

I also discovered `randombool()` and `randombool_seeded() always return false because `numerator.Value/denominator.Value` is always 0 due to floor division, so I fixed that too.

Finally, I fixed several calls to the deprecated [rand.Seed()](https://pkg.go.dev/math/rand#Seed) function.

ysugimoto reviewed (approved):
nits:

From our policy, normally we don't use snake case variable name so 'randValue' or 'rv' will be fine for us, tiny thing!

MasonM commented:
@ysugimoto Okay, done in 64deaac58377f480606f30017486389ecea2cdab

richardmarshall commented:
Since you're already touching the rand functions and their valid ranges... would you mind fixing them to work with `0 -> 0` ranges?

Currently that causes a panic in falco.

`main.vcl`
```
backend foo {
	.host = "localhost";
}

sub vcl_recv {
	declare local var.randint1 INTEGER;
	declare local var.randint2 INTEGER;
	declare local var.randbool1 BOOL;
	declare local var.randbool2 BOOL;

	set var.randint1 = randomint(0,0);
	set var.randint2 = randomint_seeded(0, 0, 0);
	set var.randbool1 = randombool(0,0);
	set var.randbool2 = randombool_seeded(0, 0, 0);

	# FASTLY RECV
}
```
`main.test.vcl`
```
sub test_rand {
	testing.call_subroutine("vcl_recv");
}
```
output of `falco test main.vcl` using your fix-rand branch (same results on main, just indicating the existing commits in this MR don't fix it):

```
falco test main.vcl
Running tests...panic: invalid argument to Int63n

goroutine 5 [running]:
math/rand.(*Rand).Int63n(0x14000033150?, 0x12a404218?)
	math/rand/rand.go:122 +0xe4
math/rand.Int63n(0x14000690b80?)
	math/rand/rand.go:443 +0x28
github.com/ysugimoto/falco/interpreter/function/builtin.Randomint(0x20?, {0x14000690b80?, 0x2, 0x102704540?})
	github.com/ysugimoto/falco/interpreter/function/builtin/randomint.go:44 +0x100
github.com/ysugimoto/falco/interpreter/function.glob..func215(0x10?, {0x14000690b80?, 0x14000032f90?, 0x1026ff780?})
	github.com/ysugimoto/falco/interpreter/function/builtin_functions.go:1085 +0x20
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessFunctionCallExpression(0x140001da300, 0x140006a6e40, 0x88?)
	github.com/ysugimoto/falco/interpreter/expression.go:226 +0x4c8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessExpression(0x14000692ef0?, {0x102e40828?, 0x140006a6e40?}, 0xc8?)
	github.com/ysugimoto/falco/interpreter/expression.go:107 +0x7f8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSetStatement(0x140001da300, 0x140006a6e10)
	github.com/ysugimoto/falco/interpreter/statement.go:156 +0x34
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessBlockStatement(0x140001da300, {0x14000630b00?, 0xc, 0x102df94c0?}, 0x102d7dc00?)
	github.com/ysugimoto/falco/interpreter/statement.go:34 +0x1b8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSubroutine(0x140001da300, 0x140006909c0, 0x14000692830?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:42 +0x284
github.com/ysugimoto/falco/tester/function.Testing_call_subroutine(0x140006a2e00, 0x102afc984?, {0x14000033120?, 0x1, 0x1?})
	github.com/ysugimoto/falco/tester/function/testing_call_subroutine.go:50 +0x100
github.com/ysugimoto/falco/tester/function.TestingFunctions.func1(0x140006a2e00, {0x14000033120?, 0x1400000e150?, 0x17?})
	github.com/ysugimoto/falco/tester/function/testing_functions.go:27 +0x38
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessFunctionCallStatement(0x140001da300, 0x140006a6690, 0x0?)
	github.com/ysugimoto/falco/interpreter/statement.go:394 +0x74c
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessBlockStatement(0x140001da300, {0x14000033110?, 0x1, 0x102afd984?}, 0x102d7dc00?)
	github.com/ysugimoto/falco/interpreter/statement.go:67 +0x58c
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSubroutine(0x140001da300, 0x140006908c0, 0x1258a96f888b5?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:42 +0x284
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessTestSubroutine(0x140001da300?, 0x140006908c0?, 0x1032ace20?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:17 +0x34
github.com/ysugimoto/falco/tester.(*Tester).run.func1(0x0?)
	github.com/ysugimoto/falco/tester/tester.go:145 +0x214
created by github.com/ysugimoto/falco/tester.(*Tester).run in goroutine 1
	github.com/ysugimoto/falco/tester/tester.go:130 +0x3a4
```

With the calls to `randint`/`randint_seeded` commented out the `randbool` functions produce a different panic
```
falco test main.vcl
Running tests...panic: runtime error: integer divide by zero

goroutine 20 [running]:
github.com/ysugimoto/falco/interpreter/function/builtin.Randombool(0x20?, {0x14000461020?, 0x2, 0x0?})
	github.com/ysugimoto/falco/interpreter/function/builtin/randombool.go:48 +0x18c
github.com/ysugimoto/falco/interpreter/function.glob..func211(0x10?, {0x14000461020?, 0x140001bf010?, 0x1000e3780?})
	github.com/ysugimoto/falco/interpreter/function/builtin_functions.go:1065 +0x20
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessFunctionCallExpression(0x14000084360, 0x1400039f3b0, 0x88?)
	github.com/ysugimoto/falco/interpreter/expression.go:226 +0x4c8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessExpression(0x14000011c80?, {0x100824828?, 0x1400039f3b0?}, 0xc8?)
	github.com/ysugimoto/falco/interpreter/expression.go:107 +0x7f8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSetStatement(0x14000084360, 0x1400039f380)
	github.com/ysugimoto/falco/interpreter/statement.go:156 +0x34
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessBlockStatement(0x14000084360, {0x14000144b00?, 0xa, 0x1007dd4c0?}, 0x100761c00?)
	github.com/ysugimoto/falco/interpreter/statement.go:34 +0x1b8
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSubroutine(0x14000084360, 0x14000460dc0, 0x140000115c0?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:42 +0x284
github.com/ysugimoto/falco/tester/function.Testing_call_subroutine(0x140000afc00, 0x1004e0984?, {0x140001bf130?, 0x1, 0x1?})
	github.com/ysugimoto/falco/tester/function/testing_call_subroutine.go:50 +0x100
github.com/ysugimoto/falco/tester/function.TestingFunctions.func1(0x140000afc00, {0x140001bf130?, 0x1400000e570?, 0x17?})
	github.com/ysugimoto/falco/tester/function/testing_functions.go:27 +0x38
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessFunctionCallStatement(0x14000084360, 0x1400039eba0, 0x0?)
	github.com/ysugimoto/falco/interpreter/statement.go:394 +0x74c
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessBlockStatement(0x14000084360, {0x140001bf120?, 0x1, 0x1004e1984?}, 0x100761c00?)
	github.com/ysugimoto/falco/interpreter/statement.go:67 +0x58c
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessSubroutine(0x14000084360, 0x14000460be0, 0x125c0ba3ed548?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:42 +0x284
github.com/ysugimoto/falco/interpreter.(*Interpreter).ProcessTestSubroutine(0x14000084360?, 0x14000460be0?, 0x100c90e20?)
	github.com/ysugimoto/falco/interpreter/subroutine.go:17 +0x34
github.com/ysugimoto/falco/tester.(*Tester).run.func1(0x0?)
	github.com/ysugimoto/falco/tester/tester.go:145 +0x214
created by github.com/ysugimoto/falco/tester.(*Tester).run in goroutine 1
	github.com/ysugimoto/falco/tester/tester.go:130 +0x3a4
```

Fiddle showing these combinations don't result in any errors in Fastly: https://fiddle.fastly.dev/fiddle/8e31a9f2

MasonM commented:
@richardmarshall Thanks for the details! I can't reproduce the crash with `randomint(0, 0)` or `randomint_seeded(0, 0, 0)`, and I added a couple test cases for that. I can reproduce it with `randombool(0, 0)` and `randombool_seeded(0, 0, 0)`, so I fixed that. I also discovered those two functions are always returning `false` because `numerator.Value/denominator.Value` is always 0 due to floor division, so I fixed that too.

richardmarshall commented:
> I can't reproduce the crash with randomint(0, 0) or randomint_seeded(0, 0, 0)
Noticed that I was accidentally using a build of falco from main when testing yesterday and not a build from your branch like I had intended. That's why I was still seeing the panics for `randint*`.

richardmarshall reviewed (approved):
LGTM
